src/set_theory/ordinal/arithmetic.lean

```diff
@@ -2400,3 +2400,42 @@ meta def positivity_opow : expr → tactic strictness
 | _ := failed
 
 end tactic
+
+namespace acc
+variables {a b : α}
+
+/-- The length of the longest `r`-chain of elements below and including `a`. -/
+noncomputable def rank (h : acc r a) : ordinal :=
+acc.rec_on h $ λ a h ih, ordinal.sup $ λ b : {b // r b a}, order.succ $ ih b b.2
+
+lemma rank_eq (h : acc r a) :
+  h.rank = ordinal.sup (λ b : {b // r b a}, order.succ (h.inv b.2).rank) :=
+by { change (acc.intro a $ λ _, h.inv).rank = _, refl }
+
+/-- if `r a b` then the rank of `a` is less than the rank of `b`. -/
+lemma rank_lt_of_rel (hb : acc r b) (h : r a b) : (hb.inv h).rank < hb.rank :=
+(order.lt_succ _).trans_le $ by { rw hb.rank_eq, refine le_trans _ (ordinal.le_sup _ ⟨a, h⟩), refl }
+
+end acc
+
+namespace well_founded
+variables (hwf : well_founded r) {a b : α}
+include hwf
+
+/-- The length of the longest `r`-chain of elements below and including in `a`. -/
```

Comment:
This docstring doesn't make much sense:

First it's not clear what a "chain" means: if it means [list.chain](https://leanprover-community.github.io/mathlib_docs/data/list/defs.html#list.chain) then it's always finite and even if you take the supremum you can at most get the ordinal ω. The other possibility is [is_chain](https://leanprover-community.github.io/mathlib_docs/order/chain.html#is_chain), but then we probably need `r` to be transitive, because {0,1,2} with 0<1<2 (but 0 unrelated to 2) doesn't form a (<)-chain, so the maximum length is only 2.

If `r` is transitive then it's the (<)-relation of a preorder, and when restricted to a chain and antisymmetrized it's a well order, so it has an ordinal, and we can interpret the "length" as the ordinal. However, even in this case the supremum of the lengths of chains may not agree with the rank: consider ℕ ∪ {∞, ∞ + 1} equipped with the partial order generated by 0<∞, 1<2<∞, 3<4<5<∞, ..., ∞<∞+1. Then ∞+1 has rank ω+1, but all chains are finite so the supremum is ω.

So the docstring is not good and I suggest we just modify the definition on [this Wikipedia page](https://en.wikipedia.org/wiki/Von_Neumann_universe):
> The rank of a well-founded set is defined inductively as the smallest ordinal number greater than the ranks of all members of the set.

```suggestion
/-- The rank of an element `a` accessible under a relation `r` is defined inductively as the
  smallest ordinal greater than the ranks of all elements below it
  (i.e. elements `b` such that `r b a`). -/
```